### PR TITLE
feat(node): Use `*EventMap` interfaces for typed `EventEmitter` methods

### PR DESCRIPTION
**Pull request by ExE-Boss**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44811#issuecomment-629842455>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by SimonSchick** [comment]
I'd much rather see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44811 land and make EventEmitter generic as it removes MUCH more boilerplate (on/off/once etc.) and is cleaner overall.

**Comment by SimonSchick** [comment]
An additional approach is to create an interface `Events<T>` that defines all the event emitter overrides and use declaration merging for each type that extends EventEmitter, eg.

```
interface Socket extends Events<{
   ...
}> {}
class Socket extends EventEmitter { }
```

**Comment by ExE-Boss** [comment]
@SimonSchick That causes several instances of TS2320:
```
Interface '...' cannot simultaneously extend types 'EventEmitter' and 'TypedEventEmitter<...>'.
  Named property 'addListener' of types 'EventEmitter' and 'TypedEventEmitter<...>' are not identical.
```

File: types/node/fs.d.ts [hunk 1]
```diff
@@ -108,38 +108,52 @@ declare module "fs" {
...
+        prependOnceListener<K extends keyof FSWatcherEventMap>(event: K, listener: FSWatcherEventMap[K]): this;
+        prependOnceListener(event: string | symbol, listener: (...args: any[]) => void): this;
+
+        removeListener<K extends keyof FSWatcherEventMap>(event: K, listener: FSWatcherEventMap[K]): this;
+        removeListener(event: string | symbol, listener: (...args: any[]) => void): this;
+
+        off<K extends keyof FSWatcherEventMap>(event: K, listener: FSWatcherEventMap[K]): this;
+        off(event: string | symbol, listener: (...args: any[]) => void): this;
+
+		listeners<K extends keyof FSWatcherEventMap>(event: K): Array<FSWatcherEventMap[K]>;
```
Comment (danger-public) [inline types/node/fs.d.ts:150]: - :no_entry_sign: The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.

**Comment by ExE-Boss** [comment]
@orta The @danger-public bot has hit the rate limit for the **GitHub API**:

<pre>Request failed [403]: <a href="https://api.github.com/repos/DefinitelyTyped/DefinitelyTyped/pulls/44857">https://api.github.com/repos/DefinitelyTyped/DefinitelyTyped/pulls/44857</a>
Response: {
	"message": "API rate limit exceeded for user ID 56922856.",
	"documentation_url": "<a href="https://developer.github.com/v3/#rate-limiting">https://developer.github.com/v3/#rate-limiting</a>"
}</pre>

<https://travis-ci.org/github/DefinitelyTyped/DefinitelyTyped/builds/688995037#L3235>

**Comment by SimonSchick** [comment]
That's odd, can you post the code you created?

**Comment by connor4312** [comment]
> make EventEmitter generic

This would be great. A similar match approach with the signature `EventEmitter<EM = { [event: string]: (...args: any[]) => void }>` should be entirely unbreaking and allow opt-in type safety in libraries and for Node.js consumers.

Not sure if I prefer a method in the map versus a tuple, since the return value of the method is irrelevant, but a tuple loses argument names.

File: types/node/child_process.d.ts [hunk 2]
```diff
@@ -7,6 +7,14 @@ declare module "child_process" {
     type Serializable = string | object | number | boolean;
     type SendHandle = net.Socket | net.Server;
 
+    interface ChildProcessEventMap {
+        "close": (code: number, signal: NodeJS.Signals) => void;
```
[the remark this answers is not in the document]
Comment (ExE-Boss) [reply types/node/child_process.d.ts:11]: That’s what [`lib.dom.d.ts`](https://github.com/microsoft/TypeScript/blob/45cf20ca70ae877aad83d767af43690d302efa51/lib/lib.dom.d.ts#L5671-L5759) uses.

**Comment by ExE-Boss** [comment]
The issue with making `EventEmitter` generic is that it breaks the tests of many of `@types/node`’s dependents, which use `// $ExpectType EventEmitter`.

**Comment by SimonSchick** [comment]
@ExE-Boss How many are there?

**Comment by ExE-Boss** [comment]
@SimonSchick

Enough to make fixing <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44811> a massive pain.

There’s also a lot of assignability errors in there.

**Comment by weswigham** [comment]
@SimonSchick so what do you think we should do here?

**Comment by RyanCavanaugh** [comment]
bad bot

**Comment by elibarzilay** [comment]
(Saying something so the bot will notice this PR...)

**Comment by ExE-Boss** [comment]
@elibarzilay Well, I’ve already fixed this PR before you opened <https://github.com/DefinitelyTyped/dt-mergebot/pull/170>, so this PR isn’t likely to break soon unless another merge conflict comes up.

**Comment by RyanCavanaugh** [comment]
@SimonSchick thoughts on the current state of the PR?

**Comment by elibarzilay** [comment]
This has been open for a long time now, and IIRC, it was related to #44811 which is draft -- so I'm changing this one to draft too.

**Comment by ExE-Boss** [comment]
@elibarzilay Unlike <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44811>, which has several compatibility issues, this PR achieves better typing of `EventEmitter` subclasses without breaking compatibility with other packages.

**Comment by elibarzilay** [comment]
Well, so far the only relevant responses I see are [a preference for the other one](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44857#issuecomment-630460952) from three people.

**Pull request closed without merging**